### PR TITLE
Add initial ACI support 

### LINKED ISSUE #7648: Serverless container runtimes
### Overview of feature request

Given the importance of serverless infrastructure in the modern application landscape, it is a priority for Radius to expand beyond Kubernetes and support serverless compute platforms. Not to mention, this feature is a common ask from potential enterprise customers who may operate tech stacks beyond Kubernetes containers. 

### Acceptance criteria

1. Genericize the Radius platform to support deployment of applications on serverless (and other) compute platforms beyond Kubernetes.
2. Enable developers to deploy and manage serverless containers in a way that is consistent with the existing Radius environment and application model.
3. Make all Radius features (Recipes, Connections, App Graph, etc.) available to engineers building applications on Radius-enabled supported serverless platforms.
4. Radius support for unopinionated or process-driven serverless container runtimes (e.g. Azure Container Instances, Azure Container Apps, AWS Elastic Container Service).

**Scenario 1: Model Radius Environment, Application, and Container resources for serverless**

Enable the ability to define and run serverless compute with necessary extensions in a Radius environment.
Radius application definition for applications running on serverless compute platforms. Must be sure to include serverless platform specific customizations via a runtimes property, similar to how Kubernetes patching was implemented for containers.

Radius abstraction for a container resource that can be deployed to serverless compute platforms, leveraging the runtimes property. Must be sure to include serverless platform specific configurations via connections, extensions, and routing to other resources via gateways.

One idea to explore is whether we can we build extensibility via Recipes - i.e. allow Recipes for Containers, which themselves can be serverless containers.

**Scenario 2: User interfaces for serverless--Radius API, CLI, Dashboard**

Enable deployment and management of serverless compute resources via the existing Radius API and CLI commands.
Serverless resources that are modeled in Radius should be available in the App Graph and Dashboard for visualization and management.

**Scenario 3: Punch-through to platform-specific features and incremental adoption of Radius into existing serverless applications**

Allow for platform-specific features to be used in Radius applications via abstraction "punch-through" mechanisms, similar to how Kubernetes-specific features are supported in Radius via base YAML or PodSpec patching functionalities. 

Stretch goal: Ability to add Radius to existing serverless applications without requiring a full rewrite of the application, similar to how Radius can be added to existing Kubernetes applications via Kubernetes manifest or Helm chart annotations. Note: The Kubernetes/Helm support works because Kubernetes itself is extensible. Other systems like ACA are not extensible in the same way but we should explore if there are options to make this work.

### Additional context

_No response_

### Would you like to support us?

- [ ] Yes, I would like to support you

[AB#12430](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/12430)

### PR DESCRIPTION
**Pull request by sk593**
# Description

This adds initial support for deploying applications to an ACI runtime. 

## Type of change

- This pull request adds or changes features of Radius and has an approved issue #7648

Fixes: #7648

## Contributor checklist
Please verify that the PR meets the following requirements, where applicable:

- An overview of proposed schema changes is included in a linked GitHub issue.
    - [X] Yes
    - [ ] Not applicable
- A design document PR is created in the [design-notes repository](https://github.com/radius-project/design-notes/), if new APIs are being introduced.
    - [X] Yes
    - [ ] Not applicable
- The design document has been reviewed and approved by Radius maintainers/approvers.
    - [X] Yes
    - [ ] Not applicable
- A PR for the [samples repository](https://github.com/radius-project/samples) is created, if existing samples are affected by the changes in this PR.
    - [ ] Yes
    - [X] Not applicable
- A PR for the [documentation repository](https://github.com/radius-project/docs) is created, if the changes in this PR affect the documentation or any user facing updates are made.
    - [X] Yes [https://github.com/radius-project/docs/pull/1434]
    - [ ] Not applicable
- A PR for the [recipes repository](https://github.com/radius-project/recipes) is created, if existing recipes are affected by the changes in this PR.
    - [ ] Yes
    - [X] Not applicable

### CODE REVIEW

**Review by brooke-hamilton — CHANGES_REQUESTED** [review]
🚀 🚀 🚀 

Initial comments:
- `make generate` is creating changes and failing the PR checks.
- Please fill out the PR template.

File: pkg/corerp/handlers/azure_cgnroups.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	// for key, val := range options.DependencyProperties {
+	// 	logger.Info("dependency property", "key", key, "value", val)
+	// }
+
+	// logger.Info("ngroup dependency property", "property", options.DependencyProperties[rpv1.LocalIDAzureCGProfile])
+	cgpID, ok := options.DependencyProperties[rpv1.LocalIDAzureCGProfile]["containerGroupProfileID"]
+	if !ok {
+		return nil, errors.New("missing dependency: a user assigned identity is required to create role assignment")
+	}
+	nGroup.Properties.ContainerGroupProfiles[0].Resource.ID = to.Ptr(cgpID)
```
Comment (brooke-hamilton) [inline pkg/corerp/handlers/azure_cgnroups.go:66]: If there is any possibility of the index 0 being out of bounds we should check the array length and return an error (because this will cause a panic if the array has no 0 element.)

File: pkg/corerp/handlers/azure_cgnroups.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+		return nil, errors.New("cannot parse container group profile")
+	}
+
+	subID := options.Resource.ID.FindScope("subscriptions")
+	resourceGroupName := options.Resource.ID.FindScope("resourceGroups")
+	if subID == "" || resourceGroupName == "" {
+		return nil, fmt.Errorf("cannot find subscription or resource group in resource ID %s", options.Resource.ID)
+	}
+
+	// for key, val := range options.DependencyProperties {
```
Comment (brooke-hamilton) [inline pkg/corerp/handlers/azure_cgnroups.go:57]: Nit: remove commented code.

File: pkg/corerp/handlers/azure_containerlb.go [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+		lbresp, err := loadBalancersClient.Get(ctx, resourceGroupName, vnetID.Name(), nil)
+		if err != nil {
+			return nil, err
+		}
+
+		updateRequired := false
+
+		found := false
+		for _, conf := range lbresp.LoadBalancer.Properties.FrontendIPConfigurations {
+			if to.String(conf.Name) == *lb.Properties.FrontendIPConfigurations[0].Name {
```
Comment (brooke-hamilton) [inline pkg/corerp/handlers/azure_containerlb.go:71]: Is there any chance that any of the `*lb.Properties.*` referenced in this cde are nil, or that the arrays do not have at least one element? If not, then this is OK.

File: pkg/corerp/handlers/azure_nsg.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+type azureNSGHandler struct {
+	arm *armauth.ArmConfig
+}
+
+func (handler *azureNSGHandler) Put(ctx context.Context, options *PutOptions) (map[string]string, error) {
+	nsg, ok := options.Resource.CreateResource.Data.(*armnetwork.SecurityGroup)
+	if !ok {
+		return nil, errors.New("cannot parse subnet")
```
Comment (brooke-hamilton) [inline pkg/corerp/handlers/azure_nsg.go:41]: Should this error refer to the NSG instead of subnet?

File: pkg/corerp/handlers/azure_nsg.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	}
+
+	subID := options.Resource.ID.FindScope("subscriptions")
+	resourceGroupName := options.Resource.ID.FindScope("resourceGroups")
+	if subID == "" || resourceGroupName == "" {
+		return nil, fmt.Errorf("cannot find subscription or resource group in resource ID %s", options.Resource.ID)
+	}
+
+	if options.Resource.LocalID == rpv1.LocalIDAzureAppGWNetworkSecurityGroup {
+		publicIP, ok := options.DependencyProperties[rpv1.LocalIDAzurePublicIP]["publicIPAddress"]
```
Comment (brooke-hamilton) [inline pkg/corerp/handlers/azure_nsg.go:51]: If "publicIPAddress" does not exist this will panic. But OK we are guaranteed to have this value.

File: pkg/corerp/handlers/azure_publicip.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+type azurePublicIPHandler struct {
+	arm *armauth.ArmConfig
+}
+
+func (handler *azurePublicIPHandler) Put(ctx context.Context, options *PutOptions) (map[string]string, error) {
+	publicIP, ok := options.Resource.CreateResource.Data.(*armnetwork.PublicIPAddress)
+	if !ok {
+		return nil, errors.New("cannot parse subnet")
```
Comment (brooke-hamilton) [inline pkg/corerp/handlers/azure_publicip.go:40]: Need error message that refers to the IP address instead of subnet.

File: pkg/corerp/handlers/azure_vnetsubnet.go [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+		return nil, err
+	}
+	subnetClient := networkClientFactory.NewSubnetsClient()
+
+	vnetName := options.Resource.ID.Truncate().Name()
+
+	orgPrefix := to.String(subnet.Properties.AddressPrefix)
+
+	var lastErr error
+	for retry := 0; retry < 5; retry++ {
```
Comment (brooke-hamilton) [inline pkg/corerp/handlers/azure_vnetsubnet.go:65]: Nit: suggest making retry count a constant for easier visibility and maintenance.

File: pkg/corerp/handlers/azure_vnetsubnet.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+			subnet.Properties.AddressPrefix = to.Ptr(subnetPrefix)
+		}
+
+		poller, err := subnetClient.BeginCreateOrUpdate(ctx, resourceGroupName, vnetName, to.String(subnet.Name), *subnet, nil)
+		if err != nil {
+			lastErr = err
+			logger.Info("failed to create subnet", "error", err, "retry", retry)
+			fmt.Printf("\n\n### failed to create subnet: %s, retry: %d, %s", err.Error(), retry, *subnet.Properties.AddressPrefix)
+			time.Sleep(60 * time.Second)
```
Comment (brooke-hamilton) [inline pkg/corerp/handlers/azure_vnetsubnet.go:111]: Nit: Sleep time here (and below) could be a constant for easier visibility and maintenance.

File: pkg/corerp/renderers/aci/gateway/render.go [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+	"github.com/radius-project/radius/pkg/corerp/renderers"
+	"github.com/radius-project/radius/pkg/resourcemodel"
+	rpv1 "github.com/radius-project/radius/pkg/rp/v1"
+	"github.com/radius-project/radius/pkg/to"
+	"github.com/radius-project/radius/pkg/ucp/resources"
+)
+
+const (
+	// hard-coded for PoC.
+	aciLocation = "West US 3"
```
Comment (brooke-hamilton) [inline pkg/corerp/renderers/aci/gateway/render.go:41]: If this is not a default, should it be hard-coded?

File: pkg/corerp/renderers/aci/gateway/render.go [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+					},
+				},
+				{
+					Name: to.Ptr("AllowPublicIPAddress"),
+					Properties: &armnetwork.SecurityRulePropertiesFormat{
+						Description:              to.Ptr("Allow traffic from public ip address."),
+						Protocol:                 to.Ptr(armnetwork.SecurityRuleProtocolTCP),
+						SourceAddressPrefix:      to.Ptr("Internet"),
+						SourcePortRange:          to.Ptr("*"),
+						DestinationAddressPrefix: to.Ptr("placeholder"),
```
Comment (brooke-hamilton) [inline pkg/corerp/renderers/aci/gateway/render.go:145]: Is "placeholder" OK here?

File: pkg/corerp/renderers/aci/manualscale/render.go [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+					// Not a Kubernetes resource
+					continue
+				}
+				o, ok := ores.CreateResource.Data.(*ngroupsclient.NGroup)
+				if !ok {
+					return renderers.RendererOutput{}, errors.New("found Kubernetes resource with non-Kubernetes payload")
+				}
+
+				if e.ManualScaling != nil && e.ManualScaling.Replicas != nil {
+					o.Properties.ElasticProfile.DesiredCount = e.ManualScaling.Replicas
```
Comment (brooke-hamilton) [inline pkg/corerp/renderers/aci/manualscale/render.go:72]: Could `ElasticProfile` ever be nil?

File: pkg/corerp/renderers/aci/manualscale/render.go [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+	}
+
+	extensions := resource.Properties.Extensions
+	for _, e := range extensions {
+		switch e.Kind {
+		case datamodel.ManualScaling:
+			for _, ores := range output.Resources {
+				resourceType := ores.GetResourceType()
+				if resourceType.Provider != resourcemodel.ProviderAzure || resourceType.Type != "Microsoft.ContainerInstance/nGroups" {
+					// Not a Kubernetes resource
```
Comment (brooke-hamilton) [inline pkg/corerp/renderers/aci/manualscale/render.go:63]:
```suggestion
					// Not an NGroup resource
```

**Review by brooke-hamilton — CHANGES_REQUESTED** [review]
👍 This is great. Most of the comments below are nits or edge cases. There are some missing tests for the handlers and renderers, but I can help with adding those if you want.

File: pkg/corerp/renderers/aci/gateway/render.go (hunk 9 above)
Comment (sk593) [reply pkg/corerp/renderers/aci/gateway/render.go]: Updated to use the resource group location for all of these instead

File: pkg/corerp/handlers/azure_cgnroups.go (hunk 1 above)
Comment (sk593) [reply pkg/corerp/handlers/azure_cgnroups.go:66]: It shouldn't be. We set up a container group profile on the NGroup in the renderer

File: pkg/corerp/renderers/aci/gateway/render.go (hunk 10 above)
Comment (sk593) [reply pkg/corerp/renderers/aci/gateway/render.go:140]: This value gets updated in the handler. I'll add a comment so it's clear

File: pkg/corerp/renderers/aci/manualscale/render.go (hunk 11 above)
Comment (sk593) [reply pkg/corerp/renderers/aci/manualscale/render.go:72]: It shouldn't be since we set up the elastic profile on the NGroup object in the renderer

**Review by willdavsmith — APPROVED** [review]

**Review by brooke-hamilton — APPROVED** [review]
🚀 🚢 ✨
